controller/AuthoringTool.php

```diff
@@ -64,27 +73,78 @@ protected function getValidatedLtiMessagePayload(): LtiMessagePayloadInterface
     }
 
     /**
-     * @throws common_exception_Error
      * @throws ActionEnforcingException
      * @throws InterruptedActionException
+     * @throws LtiException
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
+     * @throws common_exception_Error
      */
     public function launch(): void
     {
-        $message = $this->getValidatedLtiMessagePayload();
+        $ltiMessage = $this->getLtiMessageOrRedirectToLogin();
 
         $user = $this->getServiceLocator()
             ->getContainer()
             ->get(tao_models_classes_UserService::class)
             ->addUser(
-                $message->getUserIdentity()->getIdentifier(),
+                $ltiMessage->getUserIdentity()->getIdentifier(),
                 helpers_Random::generateString(UserService::PASSWORD_LENGTH),
-                new core_kernel_classes_Resource(current($message->getRoles()))
+                new core_kernel_classes_Resource(current($ltiMessage->getRoles()))
             );
         $this->getServiceLocator()
             ->getContainer()
             ->get(LtiService::class)
-            ->startLti1p3Session($message, $user);
+            ->startLti1p3Session($ltiMessage, $user);
 
         $this->forward('run', null, null, $_GET);
     }
+
+    /**
+     * @return bool
```

Comment:
```suggestion
```

Reply:
I am removing it, but just to let my opinion here. I do not see the gain of this, especially using IDEs to autocomplete it, I will have to care about customizing my IDE to change its behavior of remove it by myself.